### PR TITLE
Move BackAndroid -> BackHandler, add Apple TV support for back nav

### PR DESCRIPTION
**Pull request by dlowder-salesforce**
Enable back navigation on Apple TV (with the remote's menu button) in code making use of BackAndroid.  The module is renamed to BackHandler.  BackAndroid is still exported to ReactNative for now, until external projects switch to using the new name for the module.  The navigation in https://github.com/react-community/react-navigation makes use of this module.

**Test plan**: Manual testing with an example app (https://github.com/dlowder-salesforce/react-nav-example).

### CODE REVIEW

**Comment by dlowder-salesforce** [comment]
@satya164 @ericvicenti This is a better implementation of back navigation for Apple TV.  One feature still missing: when the user is on the bottom of the nav stack, and presses the menu button, Apple TV guidelines require the app to go to background and the tvOS home screen to show.  That will require some native changes and a separate PR.

**Comment by ericvicenti** [comment]
Overall, this seems like a great direction to move in. This is a platform-agnostic concept.

The test plan doesn't mention testing on non-tv platforms, so we'd need to see that tested.

Can we warn people when they use `BackAndroid`, telling them about the breaking change?

Also the documentation doesn't seem to be affected, but we need to document this change very carefully.

**Comment by dlowder-salesforce** [comment]
The module is just a stub on iOS - the new functionality is only on tvOS.  The Android module is unchanged.

I'd have to think about how we would detect and generate a warning when an application imports BackAndroid instead of BackHandler.

Agreed on the documentation.

**Comment by ericvicenti** [comment]
For the warning, we can have a BackAndroid as a wrapper module around BackHandler, that warns whenever anything on it is called.

File: Libraries/Utilities/BackHandler.ios.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * On Apple TV, this implements back navigation using the TV remote's menu button.
+ * On iOS, this just implements a stub.
+ *
+ * @providesModule BackHandler
+ */
+
+'use strict';
+
+var Platform = require('Platform');
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: `const`

File: Libraries/Utilities/BackHandler.ios.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+var TVEventHandler = require('TVEventHandler');
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: `const`

File: Libraries/Utilities/BackHandler.ios.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+type BackPressEventName = $Enum<{
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: Wondering why it's not `type BackPressEventName = 'backPress'` instead

Also seems all examples say that the event name is `hardwareBackPress`, while the type says `backPress`

File: Libraries/Utilities/BackHandler.ios.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ *  if (!this.onMainScreen()) {
+ *    this.goBack();
+ *    return true;
+ *  }
+ *  return false;
+ * });
+ * ```
+ */
+var BackHandler;
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: `let`

File: Libraries/Utilities/BackHandler.ios.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ *  // Typically you would use the navigator here to go to the last state.
+ *
+ *  if (!this.onMainScreen()) {
+ *    this.goBack();
+ *    return true;
+ *  }
+ *  return false;
+ * });
+ * ```
+ */
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: new line after comment

File: Libraries/Utilities/BackHandler.ios.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+var BackHandler;
+
+if (Platform.isTVOS) {
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new line

File: Libraries/Utilities/BackHandler.ios.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  var _tvEventHandler = new TVEventHandler();
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: `const`

File: Libraries/Utilities/BackHandler.ios.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new line

File: Libraries/Utilities/BackHandler.ios.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      if (invokeDefault) {
+        BackHandler.exitApp();
+      }
+    }
+  });
+
+  BackHandler = {
+    exitApp: emptyFunction,
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new line

File: Libraries/Utilities/BackHandler.ios.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    addEventListener: function (
+      eventName: BackPressEventName,
+      handler: Function
+    ): {remove: () => void} {
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: The return type should be extracted out to the top,

```js
type BackHandlerSubscription = {
  remove: () => void
}
```

File: Libraries/Utilities/BackHandler.ios.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+      _backPressSubscriptions.add(handler);
+      return {
+        remove: () => BackHandler.removeEventListener(eventName, handler),
+      };
+    },
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new line

File: Libraries/Utilities/BackHandler.ios.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    removeEventListener: function(
+      eventName: BackPressEventName,
+      handler: Function
+    ): void {
+      _backPressSubscriptions.delete(handler);
+    },
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new line

File: Libraries/Utilities/BackHandler.ios.js [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  };
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new lines

File: Libraries/Utilities/BackHandler.ios.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+} else {
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new lines

File: Libraries/Utilities/BackHandler.ios.js [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+  BackHandler = {
+    exitApp: emptyFunction,
+    addEventListener() {
+      return {
+        remove: emptyFunction,
+      };
+    },
+    removeEventListener: emptyFunction,
+  };
+
```
Comment (satya164) [inline Libraries/Utilities/BackHandler.ios.js]: nit: unnecessary new lines

File: Libraries/Utilities/BackHandler.ios.js (hunk 3 above)
Comment (dlowder-salesforce) [inline Libraries/Utilities/BackHandler.ios.js]: That's copied from the Android version.

File: Libraries/Utilities/BackHandler.ios.js (hunk 9 above)
Comment (dlowder-salesforce) [inline Libraries/Utilities/BackHandler.ios.js]: I think this makes the code more readable.

File: Libraries/Utilities/BackHandler.ios.js (hunk 10 above)
Comment (dlowder-salesforce) [inline Libraries/Utilities/BackHandler.ios.js]: I'm leaving this as is because it's copied from the existing code in the previous `BackAndroid.android.js`

**Comment by dlowder-salesforce** [comment]
@sahrens Correct, RefreshControl doesn't have native support on tvOS.  Looks like the least intrusive solution is to do the platform check in ScrollView, which automatically takes care of all the other components like VirtualizedList that make use of scroll views.

**Comment by rozele** [comment]
FYI, this will also be a great thing for `react-native-windows`, since we also use the `BackAndroid` API.

File: website/server/docsList.js [hunk 16]
```diff
@@ -59,7 +59,9 @@ const apis = [
   '../Libraries/ReactNative/AppRegistry.js',
   '../Libraries/AppState/AppState.js',
   '../Libraries/Storage/AsyncStorage.js',
-  '../Libraries/Utilities/BackAndroid.android.js',
+  '../Libraries/Utilities/BackAndroid.js',
+  '../Libraries/Utilities/BackHandler.ios.js',
+  '../Libraries/Utilities/BackHandler.android.js',
```
Comment (ericvicenti) [inline website/server/docsList.js]: Does this cause there to be 3 separate pages? Lets consolidate it so that all the `BackHandler` docs are on one page.

**Comment by ericvicenti** [comment]
Also I think the RefreshControl change should be done separately

File: website/server/docsList.js (hunk 16 above)
Comment (dlowder-salesforce) [inline website/server/docsList.js]: There were only two pages when I tested this.

**Comment by dlowder-salesforce** [comment]
Ok I'll back out the RefreshControl change and submit as a separate PR

File: website/server/docsList.js (hunk 16 above)
Comment (ericvicenti) [inline website/server/docsList.js]: Ok, fair enough

**Pull request closed without merging**